cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @ingroup     drivers_periph_freqm
+ * @{
+ *
+ * @file
+ * @brief       Frequency meter driver implementation
+ *
+ * @author      Urs Gompper <urs.gompper@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "periph/freqm.h"
+
+static void _freqm_clock_init(uint8_t msr_clk_src, uint8_t ref_clk_src)
+{
+    /* Selection of the Generator and write Lock for FREQM_MSR */
+    GCLK->PCHCTRL[5].reg = GCLK_PCHCTRL_GEN(msr_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[5].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
+
+    /* Selection of the Generator and write Lock for FREQM_REF */
+    GCLK->PCHCTRL[6].reg = GCLK_PCHCTRL_GEN(ref_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[6].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
+}
+
+typedef struct {
+    freqm_callback_t callback;
+    uintptr_t context;
+} freqm_object_t;
+
+static volatile freqm_object_t freqm_obj;
+
+static void _freqm_enable(uint8_t refnum)
+{
+    FREQM->CFGA.reg = (uint16_t)(FREQM_CFGA_REFNUM(refnum));
+
+    /* Enable DONE Interrupt */
+    FREQM->INTENSET.reg = (uint8_t)FREQM_INTENSET_DONE_Msk;
```

Comment:
If the callback is only for the done interrupt, there are better ways to handle that than making it the API user's problem:

 -  if the measurement time is really short, it's fine to busy wait on the `DONE` flag.
 -  if the measurement may take some time, better make the function block by doing a `mutex_lock(&sync_lock);` after starting the measurement. The interrupt handler will then do a `mutex_unlock(&sync_lock);`

If you want periodic measurements, a callback handler that gets called each time a measurement is ready would be handy though. (But I'd still like to have a separate function for simple one-shot measurements)

Reply:
> If the callback is only for the done interrupt, there are better ways to handle that than making it the API user's problem:
> 
>    * if the measurement time is really short, it's fine to busy wait on the `DONE` flag.
> 
>    * if the measurement may take some time, better make the function block by doing a `mutex_lock(&sync_lock);` after starting the measurement. The interrupt handler will then do a `mutex_unlock(&sync_lock);`

I'm fine doing either, just have a hard time determining what *really short* is. The fact that one can drive both clocks from I/O Pins (with possibly really slow clocks) makes me realize that hypothetically this can take some time. So I guess the mutex-solution it is.

> If you want periodic measurements, a callback handler that gets called each time a measurement is ready would be handy though. (But I'd still like to have a separate function for simple one-shot measurements)

Agree

Reply:
> just have a hard time determining what really short is.

basically, shorter than the time it takes to lock a mutex, save thread context and switch to a different thread and do some work there, save context again and switch back - so *really* short (a few µs at most).

I imagine measuring a frequency takes longer than that, but if you are curious you can of course measure it yourself.